[Reporting API] Hook up to Content-Security-Policy 'report-to' directive

brentfulgham opened the pull request:
#### 9c7dc46c0bbd448e1ac26e259cf6210a6fd202f5
<pre>
[Reporting API] Hook up to Content-Security-Policy &apos;report-to&apos; directive
<a href="https://bugs.webkit.org/show_bug.cgi?id=244143">https://bugs.webkit.org/show_bug.cgi?id=244143</a>
&lt;rdar://problem/98900892&gt;

Reviewed by NOBODY (OOPS!).

Implement the stubs from Bug 243908 so that we can generate Reporting API reports for Content Security Policy rules.

This patch does the following:

1. Adds support for the &quot;Reporting-Endpoints&quot; header.
2. Exposes a Reporting object on Document and WorkerGlobalScope that can handle reports.
3. Implements a CSSViolationReportBody class (and JS binding)
4. Update the ContentSecurityPolicy implementation to generate a CSSViolationReportBody when the CSP directives include a report-to declaration.
5. Revises the CSP implementation to construct the JSON report body to match the new specification.
</pre>

rniwa reviewed (approved):
r=me

weinig commented on Source/WebCore/Modules/reporting/ReportingObserverCallback.idl:
```diff
@@ -0,0 +1,31 @@
+ 
+// https://w3c.github.io/reporting/#reportingobserver
+
+// https://bugs.webkit.org/show_bug.cgi?id=232548 This shouldn't need to be here.
```
Probably not necessary to put these typedef related comments. It is pretty well understood that typedefs are local to the IDL file in our implementation and if we had this comment for each one it would be way too much noise.

brentfulgham replied:
Thank you for taking a look! I will remove the comments before landing.

brentfulgham replied:
Removed in current version of the patch.

brentfulgham committed:
Implement stubs for Reporting API
https://bugs.webkit.org/show_bug.cgi?id=243908
<rdar://problem/98594486>

Reviewed by Sam Weinig and Ryosuke Niwa.

This implements a set of stubs for the Reporting API, updates our copy of the WPT reporting tests to current revision, and updates our results to reflect the new results with these changes.

WPT was synced with https://github.com/web-platform-tests/wpt/tree/33bbf4ec8b807077ec61df829cffa86254a1f954

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/reporting/README.md: Added.
* LayoutTests/imported/w3c/web-platform-tests/reporting/document-reporting-default-endpoint.https.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/reporting/document-reporting-path-absolute.https.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/reporting/generateTestReport-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/reporting/idlharness.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/reporting/idlharness.any.worker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/reporting/path-absolute-endpoint.https.sub-expected.txt: Removed.
* LayoutTests/imported/w3c/web-platform-tests/reporting/path-absolute-endpoint.https.sub.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/reporting/path-absolute-endpoint.https.sub.html.sub.headers: Removed.
* LayoutTests/imported/w3c/web-platform-tests/reporting/resources/generate-report-once.py:
* LayoutTests/imported/w3c/web-platform-tests/reporting/resources/generate-report.https.sub.html:
* LayoutTests/imported/w3c/web-platform-tests/reporting/w3c-import.log:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/reporting/Report.cpp: Added.
(WebCore::Report::create):
(WebCore::Report::Report):
* Source/WebCore/Modules/reporting/Report.h: Added.
(WebCore::Report::type const):
(WebCore::Report::url const):
(WebCore::Report::body):
* Source/WebCore/Modules/reporting/Report.idl: Added.
* Source/WebCore/Modules/reporting/ReportBody.h: Added.
(WebCore::ReportBody::create):
* Source/WebCore/Modules/reporting/ReportBody.idl: Added.
* Source/WebCore/Modules/reporting/ReportingObserver.cpp: Added.
(WebCore::ReportingObserver::create):
(WebCore::ReportingObserver::ReportingObserver):
(WebCore::ReportingObserver::~ReportingObserver):
(WebCore::ReportingObserver::disconnect):
(WebCore::ReportingObserver::observe):
(WebCore::ReportingObserver::takeRecords):
* Source/WebCore/Modules/reporting/ReportingObserver.h: Added.
* Source/WebCore/Modules/reporting/ReportingObserver.idl: Added.
* Source/WebCore/Modules/reporting/ReportingObserverCallback.h: Added.
* Source/WebCore/Modules/reporting/ReportingObserverCallback.idl: Added.
* Source/WebCore/Modules/reporting/ReportingObserverOptions.idl: Added.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSReportBodyCustom.cpp: Added.
(WebCore::JSReportBody::toJSON):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/features.json:

Canonical link: https://commits.webkit.org/253443@main
